src/github.com/matrix-org/dendrite/mediaapi/writers/download.go

```diff
@@ -0,0 +1,525 @@
+		MediaMetadata: &types.MediaMetadata{
+			MediaID: mediaID,
+			Origin:  origin,
+		},
+		Logger: util.GetLogger(req.Context()),
+	}
+
+	// request validation
+	if req.Method != "GET" {
+		r.jsonErrorResponse(w, util.JSONResponse{
+			Code: 405,
+			JSON: jsonerror.Unknown("request method must be GET"),
+		})
+		return
+	}
+
+	if resErr := r.Validate(); resErr != nil {
+		r.jsonErrorResponse(w, *resErr)
+		return
+	}
+
+	// check if we have a record of the media in our database
+	mediaMetadata, err := db.GetMediaMetadata(r.MediaMetadata.MediaID, r.MediaMetadata.Origin)
+
+	if err == nil {
+		// If we have a record, we can respond from the local file
+		r.MediaMetadata = mediaMetadata
+		r.respondFromLocalFile(w, cfg.AbsBasePath)
+		return
+	} else if err == sql.ErrNoRows && r.MediaMetadata.Origin != cfg.ServerName {
+		// If we do not have a record and the origin is remote, we need to fetch it and respond with that file
+		// The following code using activeRemoteRequests is avoiding duplication of fetches from the remote server in the case
+		// of multiple simultaneous incoming requests for the same remote file - it will be downloaded once, cached and served
+		// to all clients.
+
+		mxcURL := "mxc://" + string(r.MediaMetadata.Origin) + "/" + string(r.MediaMetadata.MediaID)
+
+		for tries := 0; ; tries++ {
+			activeRemoteRequests.Lock()
+			mediaMetadata, err = db.GetMediaMetadata(r.MediaMetadata.MediaID, r.MediaMetadata.Origin)
+			if err == nil {
+				// If we have a record, we can respond from the local file
+				r.MediaMetadata = mediaMetadata
+				r.respondFromLocalFile(w, cfg.AbsBasePath)
+				activeRemoteRequests.Unlock()
+				return
+			}
+			if activeRemoteRequestCondition, ok := activeRemoteRequests.Set[mxcURL]; ok {
+				if tries >= nTries {
+					r.Logger.WithFields(log.Fields{
+						"MediaID": r.MediaMetadata.MediaID,
+						"Origin":  r.MediaMetadata.Origin,
+					}).Warn("Other goroutines are trying to download the remote file and failing.")
+					r.jsonErrorResponse(w, util.JSONResponse{
+						Code: 500,
+						JSON: jsonerror.Unknown(fmt.Sprintf("File with media ID %q could not be downloaded from %q", r.MediaMetadata.MediaID, r.MediaMetadata.Origin)),
+					})
+					activeRemoteRequests.Unlock()
+					return
+				}
+				r.Logger.WithFields(log.Fields{
+					"Origin":  r.MediaMetadata.Origin,
+					"MediaID": r.MediaMetadata.MediaID,
+				}).Info("Waiting for another goroutine to fetch the remote file.")
+				activeRemoteRequestCondition.Wait()
+				activeRemoteRequests.Unlock()
+			} else {
+				activeRemoteRequests.Set[mxcURL] = &sync.Cond{L: activeRemoteRequests}
+				activeRemoteRequests.Unlock()
+				break
+			}
+		}
+
+		r.respondFromRemoteFile(w, cfg.AbsBasePath, cfg.MaxFileSizeBytes, db, activeRemoteRequests)
+	} else {
+		// If we do not have a record and the origin is local, or if we have another error from the database, the file is not found
+		r.Logger.WithError(err).Warn("Failed to look up file in database")
```

Comment:
Please do not treat these as the same thing. If we don't have a record of the media then we just 404 and it's not something we really need to look into since randoms hit random HTTP paths all the time. If we get a database-level error, that's something we need to be whining about loudly, and should really be 500ing not telling people it doesn't exist (we couldn't check!).